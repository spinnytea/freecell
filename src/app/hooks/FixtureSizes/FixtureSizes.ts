import { scale_height } from '@/app/components/cards/CardImage';
import { CardLocation, CardSequence, Rank, RankList } from '@/app/game/card';

// IDEA (mobile) use game max cascade.length to influence TABLEAU_CARD_SPACING?
// IDEA (hud) bad layout idea: free cells on left (top down), foundation on right (top down)
//  - so tableau can start at the top of the screen?
//  - it's a layout
export const DEFAULT_CLIENT_WIDTH = 800;
export const DEFAULT_CLIENT_HEIGHT = 600;

/*
	Spacing around cards is all in percentages.
	The cards will by scaled to fit the screen, so too must the negative space.

	LR spacing is expressed in % of card width
	TB spacing is expressed in % of card height
	i.e. TB_CASCADE_OFFSET will always "just barely show the rank" because the svg is also scaled and of fixed layout
*/
const _LR_HOME_MARGIN = 1;
const _LR_HOME_GAP = 2;
const TB_HOME_TOP = 0.2;
const LR_HOME_CARD_SPACING = 1 / 6; // TODO (techdebt) simplify the math below
const TB_TABLEAU_TOP = 0.3;
const LR_TABLEAU_CARD_SPACING = 2 / 7; // TODO (techdebt) simplify the math below
const TB_CASCADE_OFFSET = 0.3;
export const PEEK_UP = 0.25;
export const PEEK_DOWN = 0.5;

export interface FixtureSizes {
	boardWidth: number;
	boardHeight: number;
	cardWidth: number;
	cardHeight: number;

	home: {
		top: number;
		cellLeft: number[];
		foundationLeft: number[];
	};

	tableau: {
		top: number;
		cascadeLeft: number[];
		offsetTop: number;
	};

	deck: {
		top: number;
		left: number;
	};
}

export function calcFixtureSizes(
	boardWidth: number = DEFAULT_CLIENT_WIDTH,
	boardHeight: number = DEFAULT_CLIENT_HEIGHT
): FixtureSizes {
	// scale LR_HOME_MARGIN based on aspect ratio
	// wider playing fields need a larger margin
	// taller playing fields can have a smaller margin
	let aspectratio = boardWidth / boardHeight;
	// smooth transition across the +/-, but negative (effectively) doubles it's scale factor
	if (aspectratio < 1) aspectratio = 1 - (1 - aspectratio) * 2;

	// minimum margin of home spacing
	const LR_HOME_MARGIN_SCALED = Math.max(LR_HOME_CARD_SPACING, _LR_HOME_MARGIN * aspectratio);
	// leave at least one space for the deck
	// leave at most 2 card widths
	// otherwise, scale down the gap with the aspect ratio
	const LR_HOME_GAP_SCALED = Math.max(
		1 + LR_HOME_CARD_SPACING * 2,
		Math.min(_LR_HOME_GAP * aspectratio, 2)
	);

	// REVIEW (mobile) scale with aspect ratio?
	// const TB_CASCADE_OFFSET_SCALED = _TB_CASCADE_OFFSET / Math.max(1, Math.min(Math.sqrt(aspectratio), 1.25));

	// cells gap foundation
	// this takes up the most space, so it determines the width of the cards
	//   lr + 4 cards (3 spaces) + gap + 4 cards (3 spaces) + lr = boardWidth
	//   spaces                                                       +  cards               = boardWidth
	// ((LR_HOME_MARGIN * 2 + LR_HOME_GAP + LR_HOME_CARD_SPACING * 6) + (1 * 8)) * cardWidth = boardWidth
	const cardWidth =
		boardWidth /
		(LR_HOME_MARGIN_SCALED * 2 + LR_HOME_GAP_SCALED + LR_HOME_CARD_SPACING * 6 + 1 * 8);

	const cardHeight = scale_height(cardWidth);

	// lr + 8 cards (7 spaces) + lr = boardWidth
	// adjust tableau margins to center cards (in px values)
	const tableauLRMargin = (boardWidth - (8 + LR_TABLEAU_CARD_SPACING * 7) * cardWidth) / 2;

	const cellLOffset = LR_HOME_MARGIN_SCALED * cardWidth;
	const foundationLOffset =
		(LR_HOME_MARGIN_SCALED + LR_HOME_CARD_SPACING * 3 + 4 + LR_HOME_GAP_SCALED) * cardWidth;

	return {
		boardWidth: toFixed(boardWidth),
		boardHeight: toFixed(boardHeight),
		cardWidth: toFixed(cardWidth),
		cardHeight: toFixed(cardHeight),

		home: {
			top: toFixed(TB_HOME_TOP * cardHeight),
			// left justified
			// < lr + cell + space + cell + …
			cellLeft: [
				toFixed((1 + LR_HOME_CARD_SPACING) * cardWidth * 0 + cellLOffset),
				toFixed((1 + LR_HOME_CARD_SPACING) * cardWidth * 1 + cellLOffset),
				toFixed((1 + LR_HOME_CARD_SPACING) * cardWidth * 2 + cellLOffset),
				toFixed((1 + LR_HOME_CARD_SPACING) * cardWidth * 3 + cellLOffset),
			],
			// left justified
			// gap << cell + space + cell + …
			foundationLeft: [
				toFixed((1 + LR_HOME_CARD_SPACING) * cardWidth * 0 + foundationLOffset),
				toFixed((1 + LR_HOME_CARD_SPACING) * cardWidth * 1 + foundationLOffset),
				toFixed((1 + LR_HOME_CARD_SPACING) * cardWidth * 2 + foundationLOffset),
				toFixed((1 + LR_HOME_CARD_SPACING) * cardWidth * 3 + foundationLOffset),
			],

			// right justified
			// … + cell + space + cell + lr >
			// foundationLeft: [
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING) * cardWidth) * 3 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING) * cardWidth) * 2 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING) * cardWidth) * 1 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING) * cardWidth) * 0 -
			// 		cellLOffset),
			// ],
		},

		tableau: {
			top: toFixed((TB_HOME_TOP + 1 + TB_TABLEAU_TOP) * cardHeight),
			// left justified
			// < lr + cell + space + cell + …
			cascadeLeft: [
				toFixed((1 + LR_TABLEAU_CARD_SPACING) * cardWidth * 0 + tableauLRMargin),
				toFixed((1 + LR_TABLEAU_CARD_SPACING) * cardWidth * 1 + tableauLRMargin),
				toFixed((1 + LR_TABLEAU_CARD_SPACING) * cardWidth * 2 + tableauLRMargin),
				toFixed((1 + LR_TABLEAU_CARD_SPACING) * cardWidth * 3 + tableauLRMargin),
				toFixed((1 + LR_TABLEAU_CARD_SPACING) * cardWidth * 4 + tableauLRMargin),
				toFixed((1 + LR_TABLEAU_CARD_SPACING) * cardWidth * 5 + tableauLRMargin),
				toFixed((1 + LR_TABLEAU_CARD_SPACING) * cardWidth * 6 + tableauLRMargin),
				toFixed((1 + LR_TABLEAU_CARD_SPACING) * cardWidth * 7 + tableauLRMargin),
			],
			offsetTop: toFixed(TB_CASCADE_OFFSET * cardHeight),
		},

		// REVIEW (hud) this is an arbitrary place to deal from
		deck: {
			top: toFixed(boardHeight - cardHeight - TB_HOME_TOP * cardHeight),
			left: toFixed(cellLOffset),
		},
	};
}

function toFixed(num: number): number {
	return parseFloat(num.toFixed(3));
}

export function calcTopLeftZ(
	fixtureSizes: FixtureSizes,
	{ fixture, data }: CardLocation,
	selection: CardSequence | null,
	rank?: Rank
): { top: number; left: number; zIndex: number; transform: string | undefined } {
	switch (fixture) {
		case 'deck':
			// TODO (animation) animate cursor (selection?) within deck
			return {
				top: fixtureSizes.deck.top,
				left: fixtureSizes.deck.left,
				zIndex: data[0],
				transform: undefined,
			};
		case 'cell':
			return {
				top: fixtureSizes.home.top,
				left: fixtureSizes.home.cellLeft[data[0]],
				zIndex: data[0],
				transform:
					selection?.location.fixture === 'cell' && selection.location.data[0] === data[0]
						? 'rotate(10deg)'
						: undefined,
			};
		case 'foundation':
			return {
				top: fixtureSizes.home.top,
				left: fixtureSizes.home.foundationLeft[data[0]],
				// XXX (techdebt) zIndex + 100 so it's above the cell/cascade
				// REVIEW (animations) cards in flight
				zIndex: (rank ? RankList.indexOf(rank) : 0) + 100,
				transform:
					selection?.location.fixture === 'foundation' && selection.location.data[0] === data[0]
						? 'rotate(10deg)'
						: undefined,
			};
		case 'cascade': {
			const ret = {
				top: fixtureSizes.tableau.top + fixtureSizes.tableau.offsetTop * data[1],
				left: fixtureSizes.tableau.cascadeLeft[data[0]],
				zIndex: data[1], // we don't really need to make one cascade strictly above another
				transform: undefined,
			};
			if (selection?.location.fixture === 'cascade' && selection.location.data[0] === data[0]) {
				const sd1 = selection.location.data[1];
				if (selection.cards.length > 1 || !selection.canMove) {
					if (data[1] > sd1) {
						ret.top += fixtureSizes.tableau.offsetTop * PEEK_DOWN;
					} else if (data[1] === sd1 && sd1 > 0) {
						ret.top -= fixtureSizes.tableau.offsetTop * PEEK_UP;
					}
				} else if (data[1] === sd1) {
					ret.top += fixtureSizes.tableau.offsetTop * PEEK_DOWN;
				}
			}
			return ret;
		}
	}
}
