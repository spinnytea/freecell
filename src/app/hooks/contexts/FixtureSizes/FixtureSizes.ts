import {
	BOTTOM_OF_CASCADE,
	CARD_FACE_CUTOFF,
	scale_height,
} from '@/app/components/cards/constants';
import { CardLocation, CardSequence, getRankForCompare, Rank } from '@/app/game/card/card';

// IDEA (hud) bad layout idea: free cells on left (top down), foundation on right (top down)
//  - so tableau can start at the top of the screen?
//  - it's a layout
export const DEFAULT_CLIENT_WIDTH = 800;
export const DEFAULT_CLIENT_HEIGHT = 600;
export type FixtureLayout = 'wide' | 'justified' | 'auto';

/*
	Spacing around cards is all in percentages.
	The cards will by scaled to fit the screen, so too must the negative space.

	LR spacing is expressed in % of card width
	TB spacing is expressed in % of card height
	i.e. TB_CASCADE_OFFSET will always "just barely show the rank" because the svg is also scaled and of fixed layout
*/
const _LR_HOME_MARGIN = 1;
const _LR_HOME_GAP = 2;
const TB_HOME_TOP = 0.2;
const _LR_HOME_CARD_SPACING = 1 / 6;
const TB_TABLEAU_TOP = 0.3;
const _LR_TABLEAU_CARD_SPACING = 2 / 7;
const TB_CASCADE_OFFSET_STANDARD = 0.2;
const TB_CASCADE_OFFSET_SMOL = 0.3;
export const PEEK_UP = 0.25;
export const PEEK_DOWN = 0.5;

export interface FixtureSizes {
	boardWidth: number;
	boardHeight: number;
	cardWidth: number;
	cardHeight: number;

	home: {
		top: number;
		cellLeft: number[];
		foundationLeft: number[];
	};

	tableau: {
		top: number;
		cascadeLeft: number[];
		offsetTop: number;
	};

	deck: {
		top: number;
		left: number;
	};
}

function times<T>(count: number, cb: (i: number) => T): T[] {
	const list: T[] = [];
	for (let i = 0; i < count; i++) list.push(cb(i));
	return list;
}

export function calcFixtureSizes({
	boardWidth,
	boardHeight,
	cellCount = 4,
	cascadeCount = 8,
	fixtureLayout = 'auto',
}: {
	boardWidth?: number;
	boardHeight?: number;
	cellCount?: number;
	cascadeCount?: number;
	fixtureLayout?: FixtureLayout;
}): FixtureSizes {
	if (!boardWidth) boardWidth = DEFAULT_CLIENT_WIDTH;
	if (!boardHeight) boardHeight = DEFAULT_CLIENT_HEIGHT;
	const foundationCount = 4;
	const homeCount = foundationCount + cellCount;
	const homeGapCount = foundationCount - 1 + cellCount - 1;
	const cascadeGapCount = cascadeCount - 1;

	// scale LR_HOME_MARGIN based on aspect ratio
	// wider playing fields need a larger margin
	// taller playing fields can have a smaller margin
	let aspectratio = boardWidth / boardHeight;

	// wide layouts try to ensure we can see the card stacks, so we need to balanace the aspect ratio
	// it's ideally targetting 16:9 or 21:9, so we can just stick with that
	if (fixtureLayout === 'wide') aspectratio = 2;
	// justified layouts will fit the width, regardless of height
	else if (fixtureLayout === 'justified') aspectratio = 0.001;
	// smooth transition across the +/-, but negative (effectively) doubles it's scale factor
	else if (aspectratio < 1) aspectratio = 1 - (1 - aspectratio) * 2.5;

	// less spacing in portait mode, more spacing in landscape
	const LR_HOME_CARD_SPACING_SCALED =
		_LR_HOME_CARD_SPACING * Math.max(0.5, Math.min(aspectratio, 2));

	// min margins of the original card spacing
	// otherwise scale up or down
	const LR_HOME_MARGIN_SCALED = Math.max(_LR_HOME_CARD_SPACING, _LR_HOME_MARGIN * aspectratio);

	// leave at least one space for the deck (…but don't actually put it there, the deal looks lame)
	// leave at most 2 card widths
	// otherwise, scale down the gap with the aspect ratio
	const LR_HOME_GAP_SCALED = Math.max(
		1 + LR_HOME_CARD_SPACING_SCALED * 2,
		Math.min(_LR_HOME_GAP * aspectratio, 2 + LR_HOME_CARD_SPACING_SCALED * 2)
	);

	// cannot be wider than the home row
	// use default spacing, or scale down to be as wide as
	const LR_TABLEAU_CARD_SPACING_SCALED = Math.min(
		_LR_TABLEAU_CARD_SPACING,
		(LR_HOME_CARD_SPACING_SCALED * homeGapCount + LR_HOME_GAP_SCALED) / cascadeGapCount
	);

	// cells gap foundation
	// this takes up the most space, so it determines the width of the cards
	//   lr + 4 cards (3 spaces) + gap + 4 cards (3 spaces) + lr = boardWidth
	//   spaces                                                       +  cards               = boardWidth
	// ((LR_HOME_MARGIN * 2 + LR_HOME_GAP + LR_HOME_CARD_SPACING_SCALED * 6) + (1 * 8)) * cardWidth = boardWidth
	const cardWidth =
		boardWidth /
		(LR_HOME_MARGIN_SCALED * 2 +
			LR_HOME_GAP_SCALED +
			LR_HOME_CARD_SPACING_SCALED * homeGapCount +
			homeCount);

	const cardHeight = scale_height(cardWidth);

	const TB_CASCADE_OFFSET =
		cardWidth < CARD_FACE_CUTOFF ? TB_CASCADE_OFFSET_SMOL : TB_CASCADE_OFFSET_STANDARD;

	// lr + 8 cards (7 spaces) + lr = boardWidth
	// adjust tableau margins to center cards (in px values)
	const tableauLRMargin =
		(boardWidth - (cascadeCount + LR_TABLEAU_CARD_SPACING_SCALED * cascadeGapCount) * cardWidth) /
		2;

	const cellLOffset = LR_HOME_MARGIN_SCALED * cardWidth;
	const foundationLOffset =
		(LR_HOME_MARGIN_SCALED +
			LR_HOME_CARD_SPACING_SCALED * (cellCount - 1) +
			cellCount +
			LR_HOME_GAP_SCALED) *
		cardWidth;

	return {
		boardWidth: toFixed(boardWidth),
		boardHeight: toFixed(boardHeight),
		cardWidth: toFixed(cardWidth),
		cardHeight: toFixed(cardHeight),

		home: {
			top: toFixed(TB_HOME_TOP * cardHeight),
			// left justified
			// < lr + cell + space + cell + …
			cellLeft: times(cellCount, (i) =>
				toFixed((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth * i + cellLOffset)
			),
			// left justified
			// gap << cell + space + cell + …
			foundationLeft: times(foundationCount, (i) =>
				toFixed((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth * i + foundationLOffset)
			),

			// right justified
			// … + cell + space + cell + lr >
			// foundationLeft: [
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth) * 3 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth) * 2 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth) * 1 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth) * 0 -
			// 		cellLOffset),
			// ],
		},

		tableau: {
			top: toFixed((TB_HOME_TOP + 1 + TB_TABLEAU_TOP) * cardHeight),
			// left justified
			// < lr + cell + space + cell + …
			cascadeLeft: times(cascadeCount, (i) =>
				toFixed((1 + LR_TABLEAU_CARD_SPACING_SCALED) * cardWidth * i + tableauLRMargin)
			),
			offsetTop: toFixed(TB_CASCADE_OFFSET * cardHeight),
		},

		deck: {
			top: toFixed(boardHeight - cardHeight - TB_HOME_TOP * cardHeight),
			left: toFixed(cellLOffset),
		},
	};
}

function toFixed(num: number): number {
	return parseFloat(num.toFixed(3));
}

export function calcTopLeftZ(
	fixtureSizes: FixtureSizes,
	{ fixture, data }: CardLocation,
	selection: CardSequence | null,
	rank?: Rank
): { top: number; left: number; zIndex: number; rotation: number } {
	switch (fixture) {
		case 'deck':
			// TODO (animation) animate cursor (selection?) within deck
			return {
				top: fixtureSizes.deck.top,
				left: fixtureSizes.deck.left,
				zIndex: data[0],
				rotation: 0,
			};
		case 'cell':
			return {
				top: fixtureSizes.home.top,
				left: fixtureSizes.home.cellLeft[data[0]],
				zIndex: data[0],
				rotation:
					selection?.location.fixture === 'cell' && selection.location.data[0] === data[0] ? 10 : 0,
			};
		case 'foundation':
			return {
				top: fixtureSizes.home.top,
				left: fixtureSizes.home.foundationLeft[data[0]],
				// zIndex + BOTTOM_OF_CASCADE so it's above the cell/cascade
				// REVIEW (animation) cards in flight
				zIndex: (rank ? getRankForCompare(rank) : 0) + BOTTOM_OF_CASCADE,
				rotation:
					selection?.location.fixture === 'foundation' && selection.location.data[0] === data[0]
						? 10
						: 0,
			};
		case 'cascade': {
			const ret = {
				top: fixtureSizes.tableau.top + fixtureSizes.tableau.offsetTop * data[1],
				left: fixtureSizes.tableau.cascadeLeft[data[0]],
				zIndex: data[1], // we don't really need to make one cascade strictly above another
				rotation: 0,
			};
			if (selection?.location.fixture === 'cascade' && selection.location.data[0] === data[0]) {
				const sd1 = selection.location.data[1];
				if (selection.cards.length > 1 || !selection.canMove) {
					if (data[1] > sd1) {
						ret.top += fixtureSizes.tableau.offsetTop * PEEK_DOWN;
					} else if (data[1] === sd1 && sd1 > 0) {
						ret.top -= fixtureSizes.tableau.offsetTop * PEEK_UP;
					}
				} else if (data[1] === sd1) {
					ret.top += fixtureSizes.tableau.offsetTop * PEEK_DOWN;
				}
			}
			return ret;
		}
	}
}
