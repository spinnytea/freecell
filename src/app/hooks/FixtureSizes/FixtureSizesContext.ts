import { createContext } from 'react';
import { FixtureSizes } from '@/app/hooks/FixtureSizes/FixtureSizes';

// initialize with dummy values
export const FixtureSizesContext = createContext<FixtureSizes>({} as FixtureSizes);
