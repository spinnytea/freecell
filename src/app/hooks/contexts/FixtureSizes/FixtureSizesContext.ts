import { createContext } from 'react';
import { FixtureSizes } from '@/app/hooks/contexts/FixtureSizes/FixtureSizes';

// initialize with dummy values
export const FixtureSizesContext = createContext<FixtureSizes>({} as FixtureSizes);
