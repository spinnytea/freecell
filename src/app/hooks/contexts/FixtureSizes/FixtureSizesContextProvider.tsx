import { MutableRefObject, ReactNode, useEffect, useState } from 'react';
import {
	calcFixtureSizes,
	DEFAULT_CLIENT_HEIGHT,
	DEFAULT_CLIENT_WIDTH,
	FixtureLayout,
	FixtureSizes,
} from '@/app/hooks/contexts/FixtureSizes/FixtureSizes';
import { FixtureSizesContext } from '@/app/hooks/contexts/FixtureSizes/FixtureSizesContext';
import { useGame } from '@/app/hooks/contexts/Game/useGame';

export function FixtureSizesContextProvider({
	gameBoardRef,
	fixtureLayout,
	children,
}: Readonly<{
	gameBoardRef: MutableRefObject<HTMLElement | null>;
	fixtureLayout?: FixtureLayout;
	children: ReactNode;
}>) {
	const game = useGame();
	const cellCount = game.cells.length;
	const cascadeCount = game.tableau.length;
	const [fixtureSizes, setFixtureSizes] = useState<FixtureSizes | null>(null);

	useEffect(() => {
		function updateSize() {
			const screenWidth = gameBoardRef.current?.offsetWidth ?? DEFAULT_CLIENT_WIDTH;
			const screenHeight = gameBoardRef.current?.offsetHeight ?? DEFAULT_CLIENT_HEIGHT;
			setFixtureSizes((fs) => {
				if (!fs || fs.boardWidth !== screenWidth || fs.boardHeight !== screenHeight) {
					return calcFixtureSizes({
						boardWidth: screenWidth,
						boardHeight: screenHeight,
						cellCount,
						cascadeCount,
						fixtureLayout,
					});
				}
				return fs;
			});
		}
		window.addEventListener('resize', updateSize);
		updateSize();
		return () => {
			window.removeEventListener('resize', updateSize);
		};
	}, [gameBoardRef, cellCount, cascadeCount, fixtureLayout]);

	// wait until we have an actual value init
	if (!fixtureSizes) {
		return null;
	}

	return (
		<FixtureSizesContext.Provider value={fixtureSizes}>{children}</FixtureSizesContext.Provider>
	);
}
