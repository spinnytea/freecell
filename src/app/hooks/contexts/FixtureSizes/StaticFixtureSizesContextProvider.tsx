import { ReactNode } from 'react';
import { FixtureSizes } from '@/app/hooks/contexts/FixtureSizes/FixtureSizes';
import { FixtureSizesContext } from '@/app/hooks/contexts/FixtureSizes/FixtureSizesContext';
import { useGame } from '@/app/hooks/contexts/Game/useGame';

export function StaticFixtureSizesContextProvider({
	children,
}: Readonly<{
	children: ReactNode;
}>) {
	const game = useGame();
	const cellCount = game.cells.length;
	const foundationCount = game.foundations.length;
	const cascadeCount = game.tableau.length;

	const fixtureSizes = calcStaticFixtureSizes(cellCount, foundationCount, cascadeCount);

	return (
		<FixtureSizesContext.Provider value={fixtureSizes}>{children}</FixtureSizesContext.Provider>
	);
}

export function calcStaticFixtureSizes(
	cellCount: number,
	foundationCount: number,
	cascadeCount: number
): FixtureSizes {
	const cellLeft = Array.from({ length: cellCount }, (_, index) => (index + 1) * 10);
	const foundationLeft = Array.from(
		{ length: foundationCount },
		(_, index) => (index + 1 + cellCount) * 10
	);
	const cascadeLeft = Array.from({ length: cascadeCount }, (_, index) => (index + 1) * 10);
	return {
		existsFixtureSizes: true,

		boardWidth: Math.max(cellCount + foundationCount, cascadeCount) * 10 + 20,
		boardHeight: 100,
		cardWidth: 10,
		cardHeight: 20,

		home: {
			top: 5,
			cellLeft,
			foundationLeft,
		},

		tableau: {
			top: 20,
			cascadeLeft,
			offsetTop: 1,
		},

		deck: {
			top: 90,
			left: 10,
		},
	};
}
