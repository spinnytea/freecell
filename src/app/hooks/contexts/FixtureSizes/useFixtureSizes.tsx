import { useContext } from 'react';
import { FixtureSizes } from '@/app/hooks/contexts/FixtureSizes/FixtureSizes';
import { FixtureSizesContext } from '@/app/hooks/contexts/FixtureSizes/FixtureSizesContext';

export function useFixtureSizes(): FixtureSizes {
	return useContext(FixtureSizesContext);
}
