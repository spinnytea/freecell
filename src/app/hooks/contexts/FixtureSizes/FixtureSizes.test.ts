import { calcCardCoords, calcFixtureSizes } from '@/app/hooks/contexts/FixtureSizes/FixtureSizes';
import { calcStaticFixtureSizes } from '@/app/hooks/contexts/FixtureSizes/StaticFixtureSizesContextProvider';
import { CardLocation } from '@/game/card/card';

describe('FixtureSizes', () => {
	describe('calcFixtureSizes', () => {
		test('basic', () => {
			expect(calcFixtureSizes({})).toMatchSnapshot();
		});

		test('portait', () => {
			expect(calcFixtureSizes({ fixtureLayout: 'portrait' })).toMatchSnapshot();
		});

		test('landscape', () => {
			expect(calcFixtureSizes({ fixtureLayout: 'landscape' })).toMatchSnapshot();
		});

		test('adjust counts', () => {
			expect(calcFixtureSizes({ cellCount: 6, cascadeCount: 10 })).toMatchSnapshot();
		});

		// the tableau can get so wide that it goes off screen
		// the assumption has always been "the home row is wider than the tableau"
		// (the home row is cell + foundation + 1) (the gap can scale down to 1 but not less)
		test.todo('cascadeCount > homeCount');
	});

	test.todo('calcTopLeftZ');

	describe('calcCardCoords', () => {
		const fixtureSizes = calcStaticFixtureSizes(4, 4, 8);
		describe('cell', () => {
			test.each`
				d0    | left
				${0}  | ${10}
				${1}  | ${20}
				${2}  | ${30}
				${3}  | ${40}
				${4}  | ${undefined}
				${-1} | ${undefined}
			`('$d0', ({ d0, left }: { d0: number; left: number }) => {
				const location: CardLocation = { fixture: 'cell', data: [d0] };
				const type = 'available-low';
				expect(calcCardCoords(fixtureSizes, location, type)).toEqual({
					top: 5,
					left,
					width: 10,
					height: 20,
				});
			});
		});

		describe('foundation', () => {
			test.each`
				d0    | left
				${0}  | ${50}
				${1}  | ${60}
				${2}  | ${70}
				${3}  | ${80}
				${4}  | ${undefined}
				${-1} | ${undefined}
			`('$d0', ({ d0, left }: { d0: number; left: number }) => {
				const location: CardLocation = { fixture: 'foundation', data: [d0] };
				const type = 'available-low';
				expect(calcCardCoords(fixtureSizes, location, type)).toEqual({
					top: 5,
					left,
					width: 10,
					height: 20,
				});
			});
		});

		describe('deck', () => {
			test.each([0, 1, 2, 3, 4, -1])('%d', (d0: number) => {
				const location: CardLocation = { fixture: 'deck', data: [d0] };
				const type = 'available-low';
				expect(calcCardCoords(fixtureSizes, location, type)).toEqual({
					top: 90,
					left: 10,
					width: 10,
					height: 20,
				});
			});
		});

		describe('cascade', () => {
			describe('single', () => {
				test.each`
					data        | top   | left
					${[0, 0]}   | ${20} | ${10}
					${[1, 1]}   | ${21} | ${20}
					${[2, 2]}   | ${22} | ${30}
					${[3, 3]}   | ${23} | ${40}
					${[4, 4]}   | ${24} | ${50}
					${[5, 5]}   | ${25} | ${60}
					${[6, 6]}   | ${26} | ${70}
					${[7, 7]}   | ${27} | ${80}
					${[8, 8]}   | ${28} | ${undefined}
					${[-1, -1]} | ${19} | ${undefined}
				`('$d0', ({ data, top, left }: { data: number[]; top: number; left: number }) => {
					const location: CardLocation = { fixture: 'cascade', data };
					const type = 'available-low';
					expect(calcCardCoords(fixtureSizes, location, type)).toEqual({
						top,
						left,
						width: 10,
						height: 20,
					});
				});
			});

			describe('column', () => {
				test.each`
					d0    | left
					${0}  | ${10}
					${1}  | ${20}
					${2}  | ${30}
					${3}  | ${40}
					${4}  | ${50}
					${5}  | ${60}
					${6}  | ${70}
					${7}  | ${80}
					${8}  | ${undefined}
					${-1} | ${undefined}
				`('$d0', ({ d0, left }: { d0: number; left: number }) => {
					const location: CardLocation = { fixture: 'cascade', data: [d0] };
					const type = 'cascade';
					expect(calcCardCoords(fixtureSizes, location, type)).toEqual({
						top: 20,
						left,
						width: 10,
						height: 119,
					});
				});
			});
		});

		describe('type = cursor', () => {
			test.each`
				location                                | top   | left
				${{ fixture: 'cell', data: [1] }}       | ${5}  | ${20}
				${{ fixture: 'foundation', data: [1] }} | ${5}  | ${60}
				${{ fixture: 'deck', data: [10] }}      | ${90} | ${10}
				${{ fixture: 'cascade', data: [0, 0] }} | ${20} | ${10}
			`('$d0', ({ location, top, left }: { location: CardLocation; top: number; left: number }) => {
				const type = 'cursor';
				expect(calcCardCoords(fixtureSizes, location, type)).toEqual({
					top,
					left,
					width: 10,
					height: 1,
				});
			});
		});
	});

	// in many cases, the top/left should be the same
	test.todo('calcTopLeftZ â¨‰ calcCardCoords');
});
