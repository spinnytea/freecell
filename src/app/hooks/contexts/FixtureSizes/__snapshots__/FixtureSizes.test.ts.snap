// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixtureSizes calcFixtureSizes adjust counts 1`] = `
{
  "boardHeight": 600,
  "boardWidth": 800,
  "cardHeight": 69,
  "cardWidth": 47.368,
  "deck": {
    "left": 63.158,
    "top": 517.2,
  },
  "home": {
    "cellLeft": [
      63.158,
      121.053,
      178.947,
      236.842,
      294.737,
      352.632,
    ],
    "foundationLeft": [
      515.789,
      573.684,
      631.579,
      689.474,
    ],
    "top": 13.8,
  },
  "tableau": {
    "cascadeLeft": [
      102.256,
      163.158,
      224.06,
      284.962,
      345.865,
      406.767,
      467.669,
      528.571,
      589.474,
      650.376,
    ],
    "offsetTop": 20.7,
    "top": 103.5,
  },
}
`;

exports[`FixtureSizes calcFixtureSizes basic 1`] = `
{
  "boardHeight": 600,
  "boardWidth": 800,
  "cardHeight": 80,
  "cardWidth": 55.385,
  "deck": {
    "left": 73.846,
    "top": 504,
  },
  "home": {
    "cellLeft": [
      73.846,
      141.538,
      209.231,
      276.923,
    ],
    "foundationLeft": [
      467.692,
      535.385,
      603.077,
      670.769,
    ],
    "top": 16,
  },
  "tableau": {
    "cascadeLeft": [
      123.077,
      194.286,
      265.495,
      336.703,
      407.912,
      479.121,
      550.33,
      621.538,
    ],
    "offsetTop": 24,
    "top": 120,
  },
}
`;

exports[`FixtureSizes calcFixtureSizes landscape 1`] = `
{
  "boardHeight": 600,
  "boardWidth": 800,
  "cardHeight": 69,
  "cardWidth": 48,
  "deck": {
    "left": 96,
    "top": 517.2,
  },
  "home": {
    "cellLeft": [
      96,
      160,
      224,
      288,
    ],
    "foundationLeft": [
      464,
      528,
      592,
      656,
    ],
    "top": 13.8,
  },
  "tableau": {
    "cascadeLeft": [
      160,
      221.714,
      283.429,
      345.143,
      406.857,
      468.571,
      530.286,
      592,
    ],
    "offsetTop": 20.7,
    "top": 103.5,
  },
}
`;

exports[`FixtureSizes calcFixtureSizes portait 1`] = `
{
  "boardHeight": 600,
  "boardWidth": 800,
  "cardHeight": 122,
  "cardWidth": 84.211,
  "deck": {
    "left": 14.035,
    "top": 441.4,
  },
  "home": {
    "cellLeft": [
      14.035,
      105.263,
      196.491,
      287.719,
    ],
    "foundationLeft": [
      428.07,
      519.298,
      610.526,
      701.754,
    ],
    "top": 36.6,
  },
  "tableau": {
    "cascadeLeft": [
      14.035,
      112.281,
      210.526,
      308.772,
      407.018,
      505.263,
      603.509,
      701.754,
    ],
    "offsetTop": 24.4,
    "top": 195.2,
  },
}
`;
